### PR TITLE
Tron signing

### PR DESCRIPTION
**Pull request by OlegGordiichuk**
Hello it is my first implementation of the network.

Could you take a look at implementation before I add unit tests ? 

Document that I was using during implementation.

https://github.com/tronprotocol/Documentation/blob/master/TRX/Tron-overview.md#103-signature

### CODE REVIEW

File: Sources/Tron/TransferAssetContract.swift [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+// This file is part of Trust. The full Trust copyright notice, including
+// terms governing use, modification, and redistribution, is contained in the
+// file LICENSE at the root of the source code distribution tree.
+
+import UIKit
+import SwiftProtobuf
+
+struct TransferAssetContract: TronTransactionContract {
+    
+    private let assetName: Data
```
Comment (vikmeup) [inline Sources/Tron/TransferAssetContract.swift]: assets name as `String`?

File: Sources/Tron/Extension/ProtocolTransaction.swift [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+//
+// This file is part of Trust. The full Trust copyright notice, including
+// terms governing use, modification, and redistribution, is contained in the
+// file LICENSE at the root of the source code distribution tree.
+
+import UIKit
+
+extension Protocol_Transaction {
+    private func hash() -> Data {
+        return Crypto.sha256(rawData.data)
```
Comment (vikmeup) [inline Sources/Tron/Extension/ProtocolTransaction.swift]: make a property.

File: Sources/Tron/Extension/Date.swift [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+// This file is part of Trust. The full Trust copyright notice, including
+// terms governing use, modification, and redistribution, is contained in the
+// file LICENSE at the root of the source code distribution tree.
+
+import UIKit
+
+extension Date {
+    var currentTimeMilliseconds: Int64 {
+        let now = timeIntervalSince1970
+        return Int64(now * 1000)
```
Comment (vikmeup) [inline Sources/Tron/Extension/Date.swift]: just `Int64(timeIntervalSince1970 * 1000)`

File: Sources/Crypto.h [hunk 4]
```diff
@@ -53,6 +53,9 @@
 /// Computes the SHA256 hash of the SHA256 hash of the data.
 + (nonnull NSData *)sha256sha256:(nonnull NSData *)data;
 
+/// Computes the SHA256 hash of the data.
++ (nonnull NSData *)sha256:(nonnull NSData *)data;
```
Comment (vikmeup) [inline Sources/Crypto.h]: there is also `hash()` which is the same thing

File: Sources/Tron/Protobuf/Tron.pb.swift [hunk 5]
```diff
@@ -0,0 +1,4581 @@
+// DO NOT EDIT.
```
Comment (vikmeup) [inline Sources/Tron/Protobuf/Tron.pb.swift]: Can you exclude files from the commit? we should run script to generate them instead

File: Sources/Crypto.h (hunk 4 above)
Comment (hewigovens) [reply Sources/Crypto.h]: @vikmeup not exactly the same, `hash()` is too general, how about rename to `keccak256` or something similar

File: Sources/Tron/Extension/Date.swift (hunk 3 above)
Comment (hewigovens) [reply Sources/Tron/Extension/Date.swift]: personal opinion, `currentTimeMilliseconds` is a bit long, like good old objc variable naming

File: Sources/Tron/Extension/ProtocolTransaction.swift [hunk 6]
```diff
@@ -0,0 +1,18 @@
+// Copyright © 2017-2018 Trust.
+//
+// This file is part of Trust. The full Trust copyright notice, including
+// terms governing use, modification, and redistribution, is contained in the
+// file LICENSE at the root of the source code distribution tree.
+
+import UIKit
+
+extension Protocol_Transaction {
```
Comment (hewigovens) [inline Sources/Tron/Extension/ProtocolTransaction.swift]: `Protocol_Transaction` is generated?

File: Sources/Tron/TronTransaction.swift [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+// This file is part of Trust. The full Trust copyright notice, including
+// terms governing use, modification, and redistribution, is contained in the
+// file LICENSE at the root of the source code distribution tree.
+
+import UIKit
+import Foundation
+
+enum TronTransactionError: Error {
+    case blockHeaderHeight
+    case blockHeaderRawData
```
Comment (hewigovens) [inline Sources/Tron/TronTransaction.swift]: maybe `LocalizedError`? block height/raw data seems not like a error

File: Sources/Tron/TronTransaction.swift [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+        let blockHeaderHash = Crypto.sha256(blockHeaderRawDataByteArray)
+        guard blockHeaderHash.count > 16 else {
+            throw TronTransactionError.blockHeaderRawData
+        }
+        let blockHash = blockHeaderHash[8...16]
+        
+        let rawData = TronTransactionRawDataBuilder()
+            .contract(contract)
+            .timestamp(Date().currentTimeMilliseconds)
+            .expiration(blockHeader.rawData.timestamp + 10 * 60 * 60 * 1000)
```
[the remark this answers is not in the document]
Comment (hewigovens) [reply Sources/Tron/TronTransaction.swift]: I'm not sure if builder pattern is suitable here, most of the properties are primitive types...

File: Sources/Tron/TronTransaction.swift (hunk 7 above)
Comment (OlegGordiichuk) [reply Sources/Tron/TronTransaction.swift]: yeah it makes sense

**Pull request closed without merging**